pkg/registry/apis/query/query.go

```diff
@@ -217,17 +262,154 @@ func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req d
 		}
 	}
 
-	// Create a response object with the error when missing (happens for client errors like 404)
-	if rsp == nil && err != nil {
-		rsp = &backend.QueryDataResponse{Responses: make(backend.Responses)}
-		for _, q := range req.Request.Queries {
-			rsp.Responses[q.RefID] = backend.DataResponse{
-				Status: backend.Status(code),
-				Error:  err,
+	return code, rsp, elapsedTime, err
+}
+
+// peformComparison determines if a comparison check should be made
+func (b *QueryAPIBuilder) peformComparison(req datasourceRequest) bool {
+	// not in passive mode
+	if b.passiveModeClient == nil {
+		return false
+	}
+
+	// do not compare relative requests - they will always be different since they
+	// are served at different times
+	_, err := strconv.ParseInt(req.Request.TimeRange.From, 10, 64)
+	_, err2 := strconv.ParseInt(req.Request.TimeRange.To, 10, 64)
+	if (err != nil) && (err2 != nil) {
+		b.log.Debug("relative query - not running comparison")
+		return false
+	}
+
+	// Only compare Loki & Prometheus requests. We do not want to compare all requests
+	// as some datasources have costs attached to queries.
+	return req.PluginId == datasources.DS_PROMETHEUS || req.PluginId == datasources.DS_LOKI
+}
+
+// queryDataAndCompare runs the query again, this time from the MT service, and compares the results. All of this happens in a go routine and is non-blocking.
+func (b *QueryAPIBuilder) queryDataAndCompare(ctx context.Context, pluginID string, pluginUID string, req datasourceRequest, stCode int, stRsp *backend.QueryDataResponse, stDuration time.Duration) {
+	client, err := (*b.passiveModeClient).GetDataSourceClient(ctx, v0alpha1.DataSourceRef{
+		Type: pluginID,
+		UID:  pluginUID,
+	})
+	if err != nil {
+		b.log.Error("unable to query", "error", err)
+		return
+	}
+
+	// the original query will likely complete before this query. Do not use the same context so the query can complete.
+	mtCode, mtRsp, elapsedTime, err := b.peformQuery(context.Background(), client, req, pluginID, pluginUID)
+	if err != nil {
+		b.increaseCompare(compareResultError, pluginID)
+		b.log.Error("unable to query", "error", err)
+		return
+	}
+
+	if mtCode != stCode {
+		b.increaseCompare(compareResultError, pluginID)
+		b.log.Error("codes do not match", "got", mtCode, "expected", stCode)
+		return
+	}
+
+	if mtRsp.Responses != nil {
+		b.compareResults(pluginID, stRsp, mtRsp)
+		// only compare duration if there is a response. Errors will usually be faster.
+		b.reportDurationDiff(pluginID, stDuration, elapsedTime)
+	}
+}
+
+func (b *QueryAPIBuilder) compareResults(pluginID string, stRsp, mtRsp *backend.QueryDataResponse) bool {
+	// If the length of both responses is different, we know that
+	// something is not right and can straight up return.
+	if len(stRsp.Responses) != len(mtRsp.Responses) {
+		b.log.Debug("response sizes are different", "expected", len(stRsp.Responses), "got", len(mtRsp.Responses))
```

Comment:
Nit. For any log message of differences might be helpful to output more details about the request to be able to see what the query was, i.e. `data.QueryDataRequest`?